src/webgpu/shader/validation/expression/call/builtin/atomics.spec.ts

```diff
@@ -0,0 +1,66 @@
+export const description = `
+Validation tests for atomic builtins.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const atomicOps = [
+  'atomicAdd(&a, 1)',
+  'atomicSub(&a, 1)',
+  'atomicMax(&a, 1)',
+  'atomicMin(&a, 1)',
+  'atomicAnd(&a, 1)',
+  'atomicOr(&a, 1)',
+  'atomicXor(&a, 1)',
+  'atomicLoad(&a)',
+  'atomicStore(&a, 1)',
+];
+
+g.test('stage')
+  .specURL('https://www.w3.org/TR/WGSL/#atomic-rmw')
+  .desc(
+    `
+Atomic built-in functions must not be used in a vertex shader stage.
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', ['fragment', 'vertex', 'compute'] as const) //
+      .combine('atomicOp', atomicOps)
+  )
+  .fn(t => {
+    let code = `
+@group(0) @binding(0) var<storage, read_write> a: atomic<i32>;
+`;
+
+    switch (t.params.stage) {
+      case 'compute':
+        code += `
+@compute @workgroup_size(1,1,1) fn main() {
+  ${t.params.atomicOp};
+}`;
+        break;
+
+      case 'fragment':
+        code += `
+@fragment fn main() -> @location(0) vec4<f32> {
+  ${t.params.atomicOp};
+  return vec4<f32>();
```

Comment:
The `vec4<f32>`'s can be `vec4f` if you wanted to shorten slightly.

Reply:
Yeah, I kinda like the explicitness here for some reason.